keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+
+Due to these complications, we have decided to evalute CEL expressions
+dynamically. Informational type checking will be provided (except for aggregated
+API server types), but will be surfaced only as warnings. See "Alternatives
+Considered" section for details of all the alternatives we reviewed when
+selecting this approach.
+
+Type checking is still performed for all expressions where a GVK can be matched
+to type check against, resulting in warnings, e.g.:
+
+```yaml
+...
+status:
+  expressionWarnings:
+    - expression: "object.foo"
+      warning: "no such field 'foo'"
+```
+
+#### Failure Policy
+
+For in-process validation there is no remote request, so errors should be
+deterministic. We believe this significantly reduces the risk of "fail closed"
+admission control as compared to webhooks.
+
+Because failure policy is most often selected based on the need to guarantee
+enforcement, we will default failure policy to "fail" and allow it to
+be configured on a per-policy basis:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+spec:
+  ...
+  failurePolicy: Ignore # The default is "Fail"
+  validations:
+    - expression: "object.spec.xyz == params.x"  
+```
+
+#### Safety measures
+
+To prevent clusters from being put into a unustable state that cannot be
+recoverd from via the API, admission webhooks are not allowed to match
+`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.
+
+We will extend this approach:
+
+- `ValidatingAdmissionPolicy` cannot match `ValidatingAdmissionPolicy`
+- `ValidatingWebhookConfiguration` cannot match `ValidatingAdmissionPolicy` or
+  `MutatingWebhookConfiguration`.
+
+Note that this does allow `ValidatingAdmissionPolicy` to match
+`ValidatingWebhookConfiguration`.
+
+Note: In the future we may further loosen this up and allow admission
+configuration to intercept/guard writes to admission configuration while
+preventing deadlock - Add feature to configure a set of webhooks to intercept
+other webhooks https://github.com/kubernetes/kubernetes/issues/101794.
+
+Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
+`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
+higher level. This could be added later.
+
+#### Default Validations
+
+A `defaultValidations` field will allow policy definition authors to define
+expressions which are evaluated when the policy definition matches a resource
+but no policy configurations match that resource.
+
+Behavior:
+
+- `defaultValidations` do not have access to params.
+- `spec.paramSource` is optional only when `defaultValidations` are present
+- (`spec.paramSource` and `spec.validations` must both be either present or absent)
+- When a policy has `defaultValidations`, it must also declare the enforcement
+  level of those validations
+
+A "singleton" (aka standalone) policy can be defined as:
```

Comment:
Probably need a "default binding" section to make this truly standalone now.

Reply:
Yes, this it out-of-date with our most recent changes. Fixing.